README.md
<div align="center" style="margin-bottom: 30px;">
<img src="https://raw.githubusercontent.com/clauderic/react-infinite-calendar/master/.github/logo.png" width="250"/>
</div>

# React Infinite Calendar
[![npm version](https://img.shields.io/npm/v/react-infinite-calendar.svg)](https://www.npmjs.com/package/react-infinite-calendar)
![downloads](https://img.shields.io/npm/dm/react-infinite-calendar.svg)
[![build](https://travis-ci.org/clauderic/react-infinite-calendar.svg?branch=master)](https://travis-ci.org/clauderic/react-infinite-calendar)
[![coverage](https://img.shields.io/codecov/c/github/clauderic/react-infinite-calendar.svg)](https://codecov.io/gh/clauderic/react-infinite-calendar)
[![bitHound Overall Score](https://www.bithound.io/github/clauderic/react-infinite-calendar/badges/score.svg)](https://www.bithound.io/github/clauderic/react-infinite-calendar)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg?maxAge=2592000)](https://github.com/clauderic/react-infinite-calendar/blob/master/LICENSE)
[![XO code style](https://img.shields.io/badge/code_style-XO-5ed9c7.svg)](https://github.com/sindresorhus/xo)
[![Gitter](https://badges.gitter.im/clauderic/react-infinite-calendar.svg)](https://gitter.im/clauderic/react-infinite-calendar)
### Examples available here: <a href="#">http://clauderic.github.io/react-infinite-calendar/</a>

Features
---------------

* **Infinite scroll** – Just keep scrollin', just keep scrollin'
* **Flexible** – Min/max date, disabled dates, disabled days, etc.
* **Localization and translation** – En français, s'il vous plaît!
* **Customizeable** – Customize and theme to your heart's content.
* **Year selection** – For rapidly jumping from year to year
* **Keyboard support** – ⬆️ ⬇️ ⬆️ ⬇️ ⬅️ ➡️ ⬅️ ➡️ ↩️
* **Events and callbacks** – beforeSelect, onSelect, onScroll, yadda yadda yadda.
* **Mobile-friendly** – Silky smooth scrolling on mobile

And mucho, mucho mas 🎉
<div style="padding:30px">
<img src="https://raw.githubusercontent.com/clauderic/react-infinite-calendar/master/.github/preview.gif" width="300" />
</div>

Installation
------------

Using [npm](https://www.npmjs.com/):

	$ npm install react-infinite-calendar --save


Then, using a module bundler that supports either CommonJS or ES2015 modules, such as [webpack](https://github.com/webpack/webpack):

```js
// Using an ES6 transpiler like Babel
import InfiniteCalendar from 'react-infinite-calendar';
import 'react-infinite-calendar/styles.css'; // Make sure to import the default stylesheet

// Not using an ES6 transpiler
var InfiniteCalendar = require('react-infinite-calendar');
require('react-infinite-calendar/styles.css');
```

Alternatively, an UMD build is also available:
```html
<link rel="stylesheet" href="react-infinite-calendar/styles.css">
<script src="react-infinite-calendar/dist/umd/react-infinite-calendar.js"></script>
```

Usage
------------
### Basic Example

```js
import React from 'react';
import { render } from 'react-dom';
import InfiniteCalendar from 'react-infinite-calendar';
import 'react-infinite-calendar/styles.css'; // only needs to be imported once

// Render the Calendar
var today = new Date();
var minDate = Number(new Date()) - (24*60*60*1000) * 7; // One week before today

render(
  <InfiniteCalendar
    width={400}
    height={600}
    selectedDate={today}
    disabledDays={[0,6]}
    minDate={minDate}
    keyboardSupport={true}
  />,
  document.getElementById('root')
);
```
For more usage examples, see [http://clauderic.github.io/react-infinite-calendar/](http://clauderic.github.io/react-infinite-calendar/)

### Prop Types
| Property             | Type                                                   | Default                                                                                              | Description                                                                                                                                                                                                                                 |
|:---------------------|:-------------------------------------------------------|:-----------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| selectedDate         | [Date](http://momentjs.com/docs/#/parsing/) or Boolean | new Date()                                                                                           | Value of the date that appears to be selected. Supports any input format supported by [moment.js](http://momentjs.com/docs/#/parsing/). Set to `false` if you don't wish to have a date initially selected.                                 |
| min                  | [Date](http://momentjs.com/docs/#/parsing/)            | new&nbsp;Date(1980,0,1)                                                                              | The minimum month that can be scrolled to. Supports any input format supported by [moment.js](http://momentjs.com/docs/#/parsing/)                                                                                                          |
| max                  | [Date](http://momentjs.com/docs/#/parsing/)            | new&nbsp;Date(2050,11,31)                                                                            | The maximum month that can be scrolled to. Supports any input format supported by [moment.js](http://momentjs.com/docs/#/parsing/)                                                                                                          |
| minDate              | [Date](http://momentjs.com/docs/#/parsing/)            | new&nbsp;Date(1980,0,1)                                                                              | The minimum date that is selectable. Supports any input format supported by [moment.js](http://momentjs.com/docs/#/parsing/)                                                                                                                |
| maxDate              | [Date](http://momentjs.com/docs/#/parsing/)            | new&nbsp;Date(2050,11,31)                                                                            | The maximum date that is selectable. Supports any input format supported by [moment.js](http://momentjs.com/docs/#/parsing/)                                                                                                                |
| display              | String                                                 | 'days'                                                                                               | Whether to display the `years` or `days` view.                                                                                                                                                                                              |
| locale               | Object                                                 | See [default locale](https://github.com/clauderic/react-infinite-calendar/blob/master/src/locale.js) | By default, React Infinite Calendar comes with the `English` locale strings. You can use this to change the language, or change the first day of the week. See [moment.js documentation](http://momentjs.com/docs/#/i18n/) for more details |
| theme                | Object                                                 | See [default theme](https://github.com/clauderic/react-infinite-calendar/blob/master/src/theme.js)   | Basic customization of the colors                                                                                                                                                                                                           |
| width                | Number or String                                       | 400                                                                                                  | Width of the calendar. Use number for pixel width, string for percentage, for example: `width={400}` or `width={'80%'}`                                                                                                                     |
| height               | Number                                                 | 600                                                                                                  | Height of the calendar                                                                                                                                                                                                                      |
| rowHeight            | Number                                                 | 56                                                                                                   | Height of each row in the calendar (each week is considered a `row`)                                                                                                                                                                        |
| className            | String                                                 |                                                                                                      | Optional CSS class name to append to the root `InfiniteCalendar` element.                                                                                                                                                                   |
| overscanMonthCount   | Number                                                 | 4                                                                                                    | Number of months to render above/below the visible months. Tweaking this can help reduce flickering during scrolling on certain browers/devices.                                                                                            |
| todayHelperRowOffset | Number                                                 | 4                                                                                                    | This controls the number of rows to scroll past before the *Today* helper appears                                                                                                                                                           |
| disabledDays         | Array of Numbers                                       |                                                                                                      | Array of days of the week that should be disabled. For example, to disable Monday and Sunday: `[0, 6]`                                                                                                                                      |
| disabledDates        | Array of [Dates](http://momentjs.com/docs/#/parsing/)  |                                                                                                      | Array of arbitrary dates that should be disabled. Supports any input format supported by [moment.js](http://momentjs.com/docs/#/parsing/). For example: `['2016-01-08', new Date(), '20160520', {year: 2015, month: 03, day: 15}]`          |
| beforeSelect         | Function                                               |                                                                                                      | Callback invoked before the state is mutated. Can be used to prevent the state from being changed by returning false. Example: `function(date) { return true / false; }`                                                                    |
| onSelect             | Function                                               |                                                                                                      | Callback invoked after beforeSelect() returns true, but before the state of the calendar updates                                                                                                                                            |
| afterSelect          | Function                                               |                                                                                                      | Callback invoked after the state of the calendar has sucessfully been updated                                                                                                                                                               |
| onScroll             | Function                                               |                                                                                                      | Callback invoked when the scroll offset changes. `function (scrollTop: number) {}`                                                                                                                                                          |
| onScrollEnd          | Function                                               |                                                                                                      | Callback invoked `150ms` after the last onScroll event is triggered. `function (scrollTop: number) {}`                                                                                                                                      |
| keyboardSupport      | Boolean                                                | true                                                                                                 | Keyboard support (`left`, `right`, `up`, `down`, `enter`)                                                                                                                                                                                   |
| autoFocus            | Boolean                                                | true                                                                                                 | Whether the Calendar root should be auto-focused when it mounts. This is useful when `keyboardSupport` is enabled (the calendar must be focused to listen for keyboard events)                                                              |
| tabIndex             | Number                                                 | 1                                                                                                    | Tab-index of the calendar                                                                                                                                                                                                                   |

CHANGELOG.md
Changelog
------------
### 1.2.2
Fixes a zIndex bug introduced with 1.2.0 with partial last month rows ([#52](https://github.com/clauderic/react-infinite-calendar/issues/52))

### 1.2.1
Hotfix release to fix borked build.

### 1.2.0
Minor refactoring tweaks to support React Virtualized 8.0.0 and up

### 1.1.18
Fixes a bug with scrolling programatically to an arbitrary value introduced with newer versions of react-virtualized ([#40](https://github.com/clauderic/react-infinite-calendar/issues/40))

### 1.1.17
Tweaking today helper behavior when scrolled above/below the fold ([#25](https://github.com/clauderic/react-infinite-calendar/issues/25))

### 1.1.16
Fixes issues with UMD build and externals.

### 1.1.15
- Fixes issues with weekdays not properly aligning with dates when scrollbars are visible ([#21](https://github.com/clauderic/react-infinite-calendar/issues/21))
- Updated to the latest version of `react-virtualized`, which addresses an edge-case bug where an initial scroll offset prop is specified before external CSS stylesheets have loaded ([#23](https://github.com/clauderic/react-infinite-calendar/issues/23))
- Fixes issues with the today helper offset calculation and added a new prop `todayHelperRowOffset` to customize the number of rows before the today helper becomes visible ([#24](https://github.com/clauderic/react-infinite-calendar/issues/24))
- Fixes issue with percentage `width` ([#14](https://github.com/clauderic/react-infinite-calendar/issues/14))

### 1.1.14
Minor bug fixes

### 1.1.13 – Year selection 🎉
Just click on the year in the header to test it out. The current year is underlined with the current `theme.todayColor` (orange by default). Keyboard support has also been tweaked to support this new functionality. Here's a preview:
<div>
<img width="325" alt="Year selection" src="https://cloud.githubusercontent.com/assets/1416436/15803422/b58e8704-2aaa-11e6-9c93-b1aa64fadc2e.png">
</div>

### 1.1.12 – Bug fixes 🙃
Fixes an issue with row calculation ([#5](https://github.com/clauderic/react-infinite-calendar/issues/5))


src/index.js
import React, {Component, PropTypes} from 'react';
import classNames from 'classnames';
import moment from 'moment';
import debounce from 'lodash/debounce';
import range from 'lodash/range';
import {getScrollSpeed, getMonthsForYear, keyCodes, parseDate, validDate, validDisplay, validLayout} from './utils';
import defaultLocale from './locale';
import defaultTheme from './theme';
import Today from './Today';
import Header from './Header';
import List from './List';
import Weekdays from './Weekdays';
import Shortcuts from './Shortcuts';
import Years from './Years';

const onClickOutside = require('react-onclickoutside');
const containerStyle = require('./Container.scss');
const expansionButtonStyle = require('./ExpansionButton.scss');
const dayStyle = require('./Day/Day.scss');
const weekStyle = require('./Week/Week.scss');
const style = {
	container: containerStyle,
	day: dayStyle,
	week: weekStyle,
	expansionButton: expansionButtonStyle,
};

class InfiniteCalendar extends Component {
	constructor(props) {
		super();

		// Initialize
		this.updateLocale(props.locale);
		this.updateYears(props);
		this.state = {
			height: props.collapsedHeight,
			expandedHeight: props.expandedHeight,
			collapsedHeight: props.collapsedHeight,
			selectedWeek: this.parseSelectedWeek(props.selectedWeek),
			selectedDate: this.parseSelectedDate(props.selectedDate),
			display: props.display,
			shouldHeaderAnimate: props.shouldHeaderAnimate,
			isCollapsed: props.isCollapsed,
			expandOnScroll: false,
			desktop: props.desktop,
			device: props.device,
		};
	}

	static defaultProps = {
		width: 400,
		expandedHeight: 400,
		collapsedHeight: 200,
		isCollapsed: true,
		rowHeight: 40,
		overscanMonthCount: 4,
		todayHelperRowOffset: 4,
		layout: 'portrait',
		display: 'days',
		selectedDate: new Date(),
		selectedWeek: null,
		min: {year: moment().subtract(2, 'years').year(), month: 0, day: 0},
		minDate: {year: moment().subtract(2, 'years').year(), month: 0, day: 0},
		max: {year: moment().add(2, 'years').year(), month: 11, day: 31},
		maxDate: {year: moment().add(2, 'years').year(), month: 11, day: 31},
		keyboardSupport: false,
		autoFocus: true,
		shouldHeaderAnimate: true,
		showOverlay: true,
		showTodayHelper: false,
		showHeader: false,
		tabIndex: 1,
		locale: {},
		theme: {},
		hideYearsOnSelect: true,
		hideYearsOnDate: true,
		showSelectionText: true,
		isClickOnDatepicker: false,
		device: true,
		desktop: true,
	};

	static propTypes = {
		selectedDate: validDate,
		selectedWeek: validDate,
		min: validDate,
		max: validDate,
		minDate: validDate,
		maxDate: validDate,
		locale: PropTypes.object,
		theme: PropTypes.object,
		width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
		expandedHeight: PropTypes.number,
		collapsedHeight: PropTypes.number,
		isCollapsed: PropTypes.bool,
		rowHeight: PropTypes.number,
		className: PropTypes.string,
		overscanMonthCount: PropTypes.number,
		todayHelperRowOffset: PropTypes.number,
		disabledDays: PropTypes.arrayOf(PropTypes.number),
		disabledDates: PropTypes.arrayOf(validDate),
		beforeSelect: PropTypes.func,
		onSelect: PropTypes.func,
		afterSelect: PropTypes.func,
		onScroll: PropTypes.func,
		onScrollEnd: PropTypes.func,
		keyboardSupport: PropTypes.bool,
		autoFocus: PropTypes.bool,
		onKeyDown: PropTypes.func,
		tabIndex: PropTypes.number,
		layout: validLayout,
		display: validDisplay,
		hideYearsOnSelect: PropTypes.bool,
		hideYearsOnDate: PropTypes.bool,
		shouldHeaderAnimate: PropTypes.bool,
		showOverlay: PropTypes.bool,
		showTodayHelper: PropTypes.bool,
		showHeader: PropTypes.bool,
		showSelectionText: PropTypes.bool,
		isClickOnDatepicker: PropTypes.bool,
		device: PropTypes.bool,
		desktop: PropTypes.bool,
	};

	componentDidMount() {
		let {autoFocus, keyboardSupport} = this.props;
		this.node = this.refs.node;
		this.list = this.refs.List;

		if (keyboardSupport && autoFocus) {
			this.node.focus();
		}
	}

	componentWillReceiveProps(next) {
		let {min, minDate, max, maxDate, locale} = this.props;
		let {display, isClickOnDatepicker, selectedWeek, selectedDate} = this.state;

		const nextDate = this.parseSelectedDate(next.selectedDate);
		const nextWeek = this.parseSelectedDate(next.selectedWeek);
		const stateDate = this.parseSelectedDate(selectedDate);
		const stateWeek = this.parseSelectedDate(selectedWeek);

		this.setState({
			collapsedHeight: next.collapsedHeight,
			expandedHeight: next.expandedHeight,
		});

		if (next.locale !== locale) {
			this.updateLocale(next.locale);
		}

		if (next.min !== min || next.minDate !== minDate || next.max !== max || next.maxDate !== maxDate) {
			this.updateYears(next);
		}

		if (next.selectedDate !== null && moment(nextDate).valueOf() !== stateDate) {
			if (new Date(nextDate).getTime() > new Date(minDate.year + "-01-01").getTime() &&
					new Date(nextDate).getTime() < new Date(maxDate.year + "-12-31").getTime()) {
				this.setState({
					selectedWeek: null,
					selectedDate: nextDate,
				});

				this.setState({
					isCollapsed: this.state.desktop,
				});

				this.clearHighlight();
				this.scrollToDate(nextDate, 0);
			}
		}
		
		if (next.selectedWeek !== null && nextWeek !== stateWeek) {
			if (new Date(nextWeek).getTime() > new Date(minDate.year + "-01-01").getTime() &&
					new Date(nextWeek).getTime() < new Date(maxDate.year + "-12-31").getTime()) {
				this.setState({
					selectedWeek: nextWeek,
					selectedDate: null,
				});

				this.setState({
					isCollapsed: this.state.desktop,
				});

				this.clearHighlight();
				this.scrollToDate(nextWeek, 0);
			}
		}

		this.setState({
			isClickOnDatepicker: false
		});

		if (next.display !== display) {
			this.setState({
				display: next.display
			});
		}
	}

	shouldComponentUpdate(nextProps, nextState) {
		// do not re-render on isTouchStarted and isScrollEnded state changes for a better scrolling experience
		const shouldUpdate = nextState.isTouchStarted === this.state.isTouchStarted && nextState.isScrollEnded === this.state.isScrollEnded;
  	return shouldUpdate;
	}

	handleClickOutside(evt) {
    if (!this.state.isCollapsed && this.state.desktop) {
   		this.setState({
   			isCollapsed: true,
   		}, () => {
   			this.clearHighlight();

				if (this.state.selectedDate !== null) {
					this.scrollToDate(this.state.selectedDate, 0);
				} else {
					this.scrollToDate(this.state.selectedWeek, 0);
				}
   		});
    }
  }

	parseSelectedDate(selectedDate) {
		if (selectedDate) {
			selectedDate = moment(selectedDate);

			// Selected Date should not be before min date or after max date
			if (selectedDate.isBefore(this._minDate)) {
				return this._minDate;
			} else if (selectedDate.isAfter(this._maxDate)) {
				return this._maxDate;
			}
		}

		return selectedDate;
	}

	parseSelectedWeek(selectedWeek) {
		if (selectedWeek) {
			selectedWeek = moment(selectedWeek);

			// Selected Date should not be before min date or after max date
			if (selectedWeek.isBefore(this._minDate)) {
				return this._minDate;
			} else if (selectedWeek.isAfter(this._maxDate)) {
				return this._maxDate;
			}
		}

		return selectedWeek;
	}

	updateYears(props = this.props) {
		let min = this._min = moment(props.min);
		let max = this._max = moment(props.max);
		this._minDate = moment(props.minDate);
		this._maxDate = moment(props.maxDate);

		this.years = range(min.year(), max.year() + 1).map((year) => getMonthsForYear(year, min, max));
		this.months = [].concat.apply([], this.years);
	}

	updateLocale(locale) {
		locale = this.getLocale(locale);
		moment.updateLocale(locale.name, locale);
		moment.locale(locale.name);
	}

	getDisabledDates(disabledDates) {
		return disabledDates && disabledDates.map((date) => moment(date).format('YYYYMMDD'));
	}

	getLocale(customLocale = this.props.locale) {
		return Object.assign({}, defaultLocale, customLocale);
	}

	getTheme(customTheme = this.props.theme) {
		return Object.assign({}, defaultTheme, customTheme);
	}

	onDaySelect = (selectedDate, e, shouldHeaderAnimate = this.props.shouldHeaderAnimate) => {
		if (selectedDate !== this.state.selectedDate) {
			let {afterSelect, beforeSelect, onSelect} = this.props;

			this.setState({
				isClickOnDatepicker: true
			});

			if (!beforeSelect || typeof beforeSelect == 'function' && beforeSelect(selectedDate)) {
				if (typeof onSelect == 'function') {
					onSelect(selectedDate, e);
				}

				const prevCollapsed = this.state.isCollapsed;

				this.setState({
					isCollapsed: this.state.desktop,
				});

				this.setState({
					selectedDate: selectedDate,
					selectedWeek: null,
				}, () => {
					this.clearHighlight();

					if (!prevCollapsed || moment(selectedDate).format('YYYYMMDD') === moment().format('YYYYMMDD')) {
						this.scrollToDate(selectedDate, 0);
					}
					
					if (typeof afterSelect == 'function') {
						afterSelect(selectedDate);
					}
				});
			}
		}
	};

	onWeekSelect = (selectedWeek) => {
		if (selectedWeek !== this.state.selectedWeek) {
			let {afterSelect, beforeSelect, onSelect} = this.props;

			if (!beforeSelect || typeof beforeSelect == 'function' && beforeSelect(selectedWeek)) {
				if (typeof onSelect == 'function') {
					onSelect(selectedWeek, e);
				}

				const prevCollapsed = this.state.isCollapsed;

				this.setState({
					isCollapsed: this.state.desktop,
				});

				this.setState({
					selectedWeek: selectedWeek,
					selectedDate: null,
					isClickOnDatepicker: true,
				}, () => {
					this.clearHighlight();

					if (!prevCollapsed || (moment(selectedWeek).format('ww') === moment().format('ww') && moment(selectedWeek).format('YYYY') === moment().format('YYYY'))) {
						this.scrollToDate(selectedWeek, 0);
					}

					if (typeof afterSelect == 'function') {
						afterSelect(selectedWeek);
					}
				});
			}
		}
	};

	getCurrentOffset = () => {
		return this.scrollTop;
	};

	getDateOffset = (date) => {
		return this.list && this.list.getDateOffset(date);
	};

	scrollTo = (offset) => {
		return this.list && this.list.scrollTo(offset);
	};

	scrollToDate = (date = moment(), offset) => {
		this.setState({
			isScrolling: false,
			expandOnScroll: false,
		}, () => {
			this.list.scrollToDate(date, offset);

			if (!this.state.isCollapsed && this.state.desktop) {
				this.setState({
					isCollapsed: true,
					expandOnScroll: true,
				});
			}
		});
	};

	handleExpansionClick = () => {
		this.setState({
			isCollapsed: false,
		});
	};

	getScrollSpeed = getScrollSpeed();

	onScroll = ({scrollTop}) => {
		let {onScroll, showOverlay, showTodayHelper, device} = this.props;
		let {isScrolling, isTouchStarted, isScrollEnded, isCollapsed, expandOnScroll} = this.state;
		let scrollSpeed = this.scrollSpeed = Math.abs(this.getScrollSpeed(scrollTop));
		this.scrollTop = scrollTop;

		if (!isScrolling && scrollSpeed > 10) {
			this.setState({
				isScrolling: true,
			});
		}

		if (isCollapsed && !device && expandOnScroll) {
			this.setState({
				isCollapsed: false,
			});
		}

		if (showTodayHelper) {
			this.updateTodayHelperPosition(scrollSpeed);
		}

		if (typeof onScroll == 'function') {
			onScroll(scrollTop);
		}

		this.onScrollEnd();
	};

	onScrollEnd = debounce(() => {
		let {onScrollEnd, showTodayHelper} = this.props;
		let {isScrolling, isTouchStarted} = this.state;

		this.setState({
			isScrollEnded: true,
			expandOnScroll: true,
		});
		
		if (isScrolling && !isTouchStarted) this.setState({isScrolling: false});
		if (showTodayHelper) this.updateTodayHelperPosition(0);
		if (typeof onScrollEnd == 'function') onScrollEnd(this.scrollTop);
	}, 150);

	handleTouchStart = () => {
		if (!this.state.isTouchStarted) {
			this.setState({
				isTouchStarted: true,
				isScrollEnded: false,
			});
		}
	};

	handleTouchEnd = () => {
		this.setState({
			isTouchStarted: false,
		});

		if (this.state.isScrollEnded) {
			this.setState({
				isScrolling: false,
			});

			this.forceUpdate();
		}
	};

	updateTodayHelperPosition = (scrollSpeed) => {
		let date = this.today.date;
		if (!this.todayOffset) this.todayOffset = this.getDateOffset(date); //scrollTop offset of the month "today" is in

		let scrollTop = this.scrollTop;
		let {showToday} = this.state;
		let {height, rowHeight, todayHelperRowOffset} = this.props;
		let newState;
		let dayOffset = Math.ceil((date.date()-7+moment(date).startOf("month").day())/7)*rowHeight; //offset of "today" within its month

		if (scrollTop >= this.todayOffset + dayOffset + rowHeight * (todayHelperRowOffset+1)) {
			if (showToday !== 1) newState = 1; //today is above the fold
		} else if (scrollTop + height <= this.todayOffset + dayOffset + rowHeight - rowHeight * (todayHelperRowOffset+1)) {
			if (showToday !== -1) newState = -1; //today is below the fold
		} else if (showToday && scrollSpeed <= 1) {
			newState = false;
		}

		if (scrollTop == 0) {
			newState = false;
		}

		if (newState != null) {
			this.setState({showToday: newState});
		}
	};

	handleKeyDown = (e) => {
		let {maxDate, minDate, onKeyDown} = this.props;
		let {display, selectedDate, highlightedDate, showToday} = this.state;
		let delta = 0;

		if (typeof onKeyDown == 'function') {
			onKeyDown(e);
		}
		if ([keyCodes.left, keyCodes.up, keyCodes.right, keyCodes.down].indexOf(e.keyCode) > -1 && typeof e.preventDefault == 'function') {
			e.preventDefault();
		}

		if (!selectedDate) {
			selectedDate = moment();
		}

		if (display == 'days') {
			if (!highlightedDate) {
				highlightedDate = selectedDate.clone();
				this.setState({highlightedDate});
			}

			switch (e.keyCode) {
				case keyCodes.enter:
					this.onDaySelect(moment(highlightedDate), e);
					return;
				case keyCodes.left:
					delta = -1;
					break;
				case keyCodes.right:
					delta = +1;
					break;
				case keyCodes.down:
					delta = +7;
					break;
				case keyCodes.up:
					delta = -7;
					break;
			}

			if (delta) {
				let {rowHeight} = this.props;
				let newHighlightedDate = moment(highlightedDate).add(delta, 'days');

				// Make sure the new highlighted date isn't before min / max
				if (newHighlightedDate.isBefore(minDate)) {
					newHighlightedDate = moment(minDate);
				} else if (newHighlightedDate.isAfter(maxDate)) {
					newHighlightedDate = moment(maxDate);
				}

				// Update the highlight indicator
				this.clearHighlight();

				// Scroll the view
				if (!this.currentOffset) this.currentOffset = this.getCurrentOffset();
				let currentOffset = this.currentOffset;
				let monthOffset = this.getDateOffset(newHighlightedDate);
				let navOffset = (showToday) ? 36 : 0;

				let highlightedEl = this.highlightedEl = this.node.querySelector(`[data-date='${newHighlightedDate.format('YYYYMMDD')}']`);

				// Edge-case: if the user tries to use the keyboard when the new highlighted date isn't rendered because it's too far off-screen
				// We need to scroll to the month of the new highlighted date so it renders
				if (!highlightedEl) {
					this.scrollTo(monthOffset - navOffset);
					return;
				}

				highlightedEl.classList.add(style.day.highlighted);

				let dateOffset = highlightedEl.offsetTop - rowHeight;
				let newOffset = monthOffset + dateOffset;


				if (currentOffset !== newOffset) {
					this.currentOffset = newOffset;
					this.scrollTo(newOffset - navOffset);
				}

				// Update the reference to the currently highlighted date
				this.setState({
					highlightedDate: newHighlightedDate
				});

			}
		} else if (display == 'years' && this.refs.years) {
			this.refs.years.handleKeyDown(e);
		}
	};

	clearHighlight() {
		if (this.highlightedEl) {
			this.highlightedEl.classList.remove(style.day.highlighted);
			this.highlightedEl = null;
		}
	}

	setDisplay = (display) => {
		this.setState({display});
	};

	render() {
		let {
			className,
			disabledDays,
			hideYearsOnSelect,
			hideYearsOnDate,
			keyboardSupport,
			layout,
			overscanMonthCount,
			min,
			minDate,
			max,
			maxDate,
			showTodayHelper,
			showHeader,
			tabIndex,
			width,
			showSelectionText,
			device,
			desktop,
			...other
		} = this.props;
		let disabledDates = this.getDisabledDates(this.props.disabledDates);
		let locale = this.getLocale();
		let theme = this.getTheme();
		let {display, isScrolling, isCollapsed, collapsedHeight, expandedHeight, selectedDate, selectedWeek, height, showToday, shouldHeaderAnimate} = this.state;
		let today = this.today = parseDate(moment());

		// Selected date should not be disabled
		if (selectedDate && (disabledDates && disabledDates.indexOf(selectedDate.format('YYYYMMDD')) !== -1 || disabledDays && disabledDays.indexOf(selectedDate.day()) !== -1)) {
			selectedDate = null;
		}

		return (
			<div 
				tabIndex={tabIndex}
				onKeyDown={keyboardSupport && this.handleKeyDown}
				className={classNames(className, style.container.root, {[style.container.landscape]: layout == 'landscape'})}
				style={{color: theme.textColor.default, width: '100%', overflow: (isCollapsed) ? 'hidden' : 'visible', height: collapsedHeight+"px" }}
				aria-label="Calendar" ref="node"
			>
				{desktop &&
					<div
						className={classNames(style.expansionButton.root, 'ion-chevron-down')}
						style={{ display: (isCollapsed) ? 'initial' : 'none'}}
						onClick={this.handleExpansionClick}
					></div>
				}
				{showHeader &&
					<Header
						selectedDate={selectedDate}
						shouldHeaderAnimate={shouldHeaderAnimate}
						layout={layout}
						theme={theme}
						locale={locale}
						scrollToDate={this.scrollToDate}
						setDisplay={this.setDisplay}
						display={display}
					/>
				}
				<div className={style.container.wrapper} >
					<Shortcuts theme={theme} locale={locale} scrollToDate={this.scrollToDate} handleTodayClick={(selectedDate !== null) ? this.onDaySelect : this.onWeekSelect} />
					<Weekdays theme={theme} locale={locale} />
					<div
						className={style.container.listWrapper}
						onTouchStart={this.handleTouchStart}
						onTouchEnd={this.handleTouchEnd}
					>
						<List
							ref="List"
							{...other}
							width={width}
							height={expandedHeight}
							selectedDate={parseDate(selectedDate)}
							selectedWeek={parseDate(selectedWeek)}
							disabledDates={disabledDates}
							disabledDays={disabledDays}
							months={this.months}
							onDaySelect={this.onDaySelect}
							onWeekSelect={this.onWeekSelect}
							onScroll={this.onScroll}
							isScrolling={isScrolling}
							today={today}
							min={parseDate(min)}
							minDate={parseDate(minDate)}
							maxDate={parseDate(maxDate)}
							theme={theme}
							locale={locale}
							overscanMonthCount={overscanMonthCount}
							showSelectionText={showSelectionText}
							hideYearsOnDate={hideYearsOnDate}
						/>
					</div>
					{display == 'years' &&
						<Years
							ref="years"
							width={width}
							height={height}
							onDaySelect={this.onDaySelect}
							minDate={minDate}
							maxDate={maxDate}
							selectedDate={selectedDate}
							theme={theme}
							years={range(moment(min).year(), moment(max).year() + 1)}
							setDisplay={this.setDisplay}
							scrollToDate={this.scrollToDate}
							hideYearsOnSelect={hideYearsOnSelect}
						/>
					}
				</div>
			</div>
		);
	}
};

export default InfiniteCalendar = onClickOutside(InfiniteCalendar);